Sources/Rendering/Core/Mapper/index.js

```diff
@@ -545,6 +552,70 @@ export function extend(publicAPI, model, initialValues = {}) {
     'scalarRange',
   ], 2);
 
+  model.viewSpecificProperties = {
+    OpenGL: {
+      ShaderReplacements: [],
+      VertexShaderCode: '',
+      FragmentShaderCode: '',
+      GeometryShaderCode: '',
+    },
+
+    addShaderReplacement(_shaderType, _originalValue, _replaceFirst, _replacementValue, _replaceAll) {
+      this.OpenGL.ShaderReplacements.push({
+        shaderType: _shaderType,
+        originalValue: _originalValue,
+        replaceFirst: _replaceFirst,
+        replacementValue: _replacementValue,
+        replaceAll: _replaceAll,
+      });
+    },
+
+    clearShaderReplacement(_shaderType, _originalValue, _replaceFirst) {
+      const shaderReplacement = this.OpenGL.ShaderReplacements;
+      let indexToRemove = -1;
+
+      for (let i = 0; i < shaderReplacement.length; i++) {
+        if (shaderReplacement[i].shaderType === _shaderType &&
+          shaderReplacement[i].originalValue === _originalValue &&
+          shaderReplacement[i].replaceFirst === _replaceFirst) {
+          indexToRemove = i;
+          break;
+        }
+      }
+      if (indexToRemove > -1) {
+        shaderReplacement.splice(indexToRemove, 1);
+      }
+    },
+
+    getShaderReplacements() {
+      return this.OpenGL.ShaderReplacements;
+    },
+
+    setVertexShaderCode(vertexShaderCode) {
+      this.OpenGL.VertexShaderCode = vertexShaderCode;
+    },
+
+    getVertexShaderCode() {
+      return this.OpenGL.VertexShaderCode;
+    },
+
+    setGeometryShaderCode(geometryShaderCode) {
+      this.OpenGL.GeometryShaderCode = geometryShaderCode;
+    },
+
+    getGeometryShaderCode() {
+      return this.OpenGL.GeometryShaderCode;
+    },
+
+    setFragmentShaderCode(fragmenShaderCode) {
+      this.OpenGL.FragmentShaderCode = fragmenShaderCode;
+    },
+
+    getFragmentShaderCode() {
+      return this.OpenGL.FragmentShaderCode;
```

Comment:
Basically I don't want `return (undefined).OpenGL.FragmentShaderCode;` to happen.